058-zk-kraft-migration.md

```diff
@@ -0,0 +1,437 @@
+The FSM is in the `ZooKeeper` state (as per `Kafka.status.metadataState` field) and it keeps staying in this state until the user applies the `strimzi.io/kraft: migration` annotation on the `Kafka` custom resource, and creates the `KafkaNodePool` custom resources(s) for the KRaft controllers.
+
+**_Trigger_**
+
+The user creates the `KafkaNodePool` custom resource(s) containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+The user sets the `strimzi.io/kraft: migration` annotation on the `Kafka` custom resource. 
+
+**_Actions_**
+
+The operator starts a reconciliation and, via the `KafkaReconciler` component, creates and configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble.
+The controllers are automatically created with the following configuration (the user cannot set them manually).
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+The operator waits for the KRaft controllers being ready, the quorum to be formed and leader elected, then the migration process moves to the state of waiting for brokers to register.
+
+**_Status changes_**
+
+The FSM moves to a `KRaftEnablingMigration` state to continue the pre-migration by configuring and rolling the brokers with the ZooKeeper migration enabled and connected to the newly created KRaft controllers.
+This state is useful if the operator crashes and it needs to recover correctly on startup, not deploying controllers but continuing with rolling brokers.
+
+---
+
+When the controllers are deployed, the operator can proceed the reconciliation by configuring and rolling brokers.
+
+**_Preconditions_**
+
+The FSM is in the `KRaftEnablingMigration` state (as per `Kafka.status.metadataState` field).
+The the `strimzi.io/kraft: migration` annotation is already applied on the `Kafka` custom resource.
+The controllers are up and running and ready with the ZooKeeper migration enabled.
+
+**_Trigger_**
+
+The operator just finished to deploy the controllers and they are ready or it's restarting after a crash while rolling the brokers.
+
+**_Actions_**
+
+During the current reconciliation and, via the `KafkaReconciler` component, the operator updates the brokers' configuration with the ZooKeeper migration flag enabled and with the new KRaft controllers quorum connection information as well.
+More specifically, the brokers are automatically updated with the following configuration (the user cannot set them manually).
+
+```shell
+# KRaft controller quorum configuration
+controller.quorum.voters=<voters>
+controller.listener.names=CONTROLLER
+
+# The controller.listener.name should also be added to listener.security.property.map
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONTROLLER:PLAINTEXT
```

Comment:
I assume if security is enabled on the controllers, the required settings are automatically included accordingly?

Reply:
Yes.